Add runtimeconfig.json support for WebAssembly

fanyang-mono opened the pull request:
Added runtimeconfig.json support for WebAssembly.
Fixes #38433

Added a test for it.
Fixes #50908

Also added a test to `Wasm.Build.Tests`, thanks to @radical

fanyang-mono committed:
Initial change for adding runtimeconfig to wasm

fanyang-mono committed:
Add runtimeconfig support to wasm

fanyang-mono committed:
Move test folder and update runtimeconifg task name

fanyang-mono committed:
Fix test

fanyang-mono committed:
Fix test csproj file

fanyang-mono committed:
Use the right var for main dll name

lambdageek commented on src/tests/FunctionalTests/WebAssembly/Browser/RuntimeConfig/WebAssembly.Browser.RuntimeConfig.Test.csproj:
```diff
@@ -0,0 +1,37 @@
+    <Copy SourceFiles="$(OutDir)\index.html" DestinationFolder="$(WasmAppDir)" />
+  </Target>
+
+  <Target Name="PreserveEnCAssembliesFromLinking"
+          Condition="'$(TargetOS)' == 'Browser' and '$(EnableAggressiveTrimming)' == 'true'"
+          BeforeTargets="ConfigureTrimming">
+    <ItemGroup>
+      <!-- Don't modify EnC test assemblies -->
+      <TrimmerRootAssembly
+          Condition="$([System.String]::Copy('%(ResolvedFileToPublish.FileName)%(ResolvedFileToPublish.Extension)').EndsWith('ApplyUpdateReferencedAssembly.dll'))"
+          Include="%(ResolvedFileToPublish.FullPath)" />
+    </ItemGroup>
+  </Target>
```
I don't think you don't need this, it's just for the hot reload test

fanyang-mono replied:
Removed.

lambdageek commented on src/tests/FunctionalTests/WebAssembly/Browser/RuntimeConfig/WebAssembly.Browser.RuntimeConfig.Test.csproj:
```diff
@@ -0,0 +1,37 @@
+    <ExpectedExitCode>42</ExpectedExitCode>
+    <WasmMainJSPath>runtime.js</WasmMainJSPath>
+    <EnableDefaultItems>false</EnableDefaultItems>
+    <!-- setting WasmXHarnessMonoArgs doesn't work here, but see runtime.js -->
+    <!-- <WasmXHarnessMonoArgs>- -setenv=DOTNET_MODIFIABLE_ASSEMBLIES=debug</WasmXHarnessMonoArgs> -->
```
These comments aren't relevant for this test.

fanyang-mono replied:
Removed.

lambdageek commented on src/tests/FunctionalTests/WebAssembly/Browser/RuntimeConfig/WebAssembly.Browser.RuntimeConfig.Test.csproj:
```diff
@@ -0,0 +1,37 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <MonoForceInterpreter>true</MonoForceInterpreter>
+    <RunAOTCompilation>false</RunAOTCompilation>
+    <PublishTrimmed>false</PublishTrimmed>
```
I don't think you need this.  your test should work with both interp and with aot.  @radical there's no need to override these unless the test wants a specific scenario, right?

fanyang-mono replied:
Removed.

lambdageek commented on src/tests/FunctionalTests/WebAssembly/Browser/RuntimeConfig/runtime.js:
```diff
@@ -0,0 +1,47 @@
+            return fetch (asset, { credentials: 'same-origin' });
+        }
+
+        if (Module.config.environment_variables !== undefined) {
+            console.log ("expected environment variables to be undefined, but they're: ", Module.config.environment_variables);
+            test_exit(1);
+        }
+        Module.config.environment_variables = {
+            "DOTNET_MODIFIABLE_ASSEMBLIES": "debug"
+        };
```
you don't need this. it's just for the hot reload test

fanyang-mono replied:
Removed.

lambdageek reviewed (commented):
The embedding changes LGTM.  There's a couple of unnecessary things copied over from the hot reload functional test that I don't think you need.

Not sure about the Wasm infrastructure changes.  @radekdoulik or @radical could oyu review those?

radical commented on src/tests/FunctionalTests/WebAssembly/Browser/RuntimeConfig/Program.cs:
```diff
@@ -0,0 +1,23 @@
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        public static int TestMeaning()
+        {
+            int result = ((string)AppContext.GetData("test_runtimeconfig_json")).Equals("25") ? 42 : 1;
```
It would be useful to print what `AppContext.GetData` returns here, for debugging.

fanyang-mono replied:
Changed.

radical commented on src/tests/FunctionalTests/WebAssembly/Browser/RuntimeConfig/WebAssembly.Browser.RuntimeConfig.Test.csproj:
```diff
@@ -0,0 +1,37 @@
+    <Scenario>WasmTestOnBrowser</Scenario>
+    <ExpectedExitCode>42</ExpectedExitCode>
+    <WasmMainJSPath>runtime.js</WasmMainJSPath>
+    <EnableDefaultItems>false</EnableDefaultItems>
```
this shouldn't be needed

fanyang-mono replied:
All of them? To my understanding, line 7-9 should be needed. But I could be wrong.

radical replied:
Setting `EnableDefaultItems=false` is useful when you want to explicitly control which source files to include, which resx files etc. But IIUC, that's not needed here.

So, if we don't set this, then even line 11(`<Compile Include="Program.cs" />`) won't be needed.

fanyang-mono replied:
`<Compile Include="Program.cs" />` is indeed needed, otherwise, I hit this error
```
CSC : error CS5001: Program does not contain a static 'Main' method suitable for an entry point [/Users/fanyang/Documents/work/dotnet_wasm/runtime/src/tests/FunctionalTests/WebAssembly/Browser/RuntimeConfig/WebAssembly.Browser.RuntimeConfig.Test.csproj]
```

radical commented on src/mono/wasm/build/WasmApp.LocalBuild.props:
```diff
@@ -34,6 +34,7 @@
     <WasmAppBuilderDir>$([MSBuild]::NormalizeDirectory('$(ArtifactsBinDir)', 'WasmAppBuilder', 'Debug', '$(_NetCoreAppToolCurrent)', 'publish'))</WasmAppBuilderDir>
     <WasmBuildTasksDir>$([MSBuild]::NormalizeDirectory('$(ArtifactsBinDir)', 'WasmBuildTasks', 'Debug', '$(_NetCoreAppToolCurrent)', 'publish'))</WasmBuildTasksDir>
     <MonoAOTCompilerDir>$([MSBuild]::NormalizeDirectory('$(ArtifactsBinDir)', 'MonoAOTCompiler', 'Debug', '$(_NetCoreAppToolCurrent)'))</MonoAOTCompilerDir>
+    <RuntimeConfigParserDir>$([MSBuild]::NormalizeDirectory('$(ArtifactsBinDir)', 'RuntimeConfigParser', 'Debug', '$(_NetCoreAppToolCurrent)'))</RuntimeConfigParserDir>
```
References to this assembly needs to be added to a few other places.

----

## Updating dependencies needed for building wasm apps

For example, if the wasm targets are using a new task, then references to that
need to be updated in a few places. Essentially, look for existing references
to `MonoAOTCompiler`, or `WasmAppBuilder` in the relevant files, and duplicate
them for the new task assembly.

1. The task assembly dir, and its path need to be in two properties:
    ```xml
    <JsonToItemsTaskFactoryDir>$([MSBuild]::NormalizeDirectory('$(ArtifactsBinDir)', 'JsonToItemsTaskFactory', 'Debug', '$(NetCoreAppToolCurrent)'))</JsonToItemsTaskFactoryDir>
    <JsonToItemsTaskFactoryTasksAssemblyPath>$([MSBuild]::NormalizePath('$(JsonToItemsTaskFactoryDir)', 'JsonToItemsTaskFactory.dll'))</JsonToItemsTaskFactoryTasksAssemblyPath>
    ```

    And this needs to be set in:
    a. `Directory.Build.props`
    b. `src/mono/wasm/build/WasmApp.LocalBuild.props`
    c. `src/tests/Common/wasm-test-runner/WasmTestRunner.proj`

2. The new dependency (eg. task assembly) needs to be sent to helix as a payload, see `src/libraries/sendtohelixhelp.proj#L365` .
3. Make changes similar to the one for existing dependent tasks in
   `eng/testing/linker/trimmingTests.targets`,
   `src/tests/Common/Directory.Build.targets#L149-L151`,
   `src/tests/Common/wasm-test-runner/WasmTestRunner.proj` and
   `src/tests/Directory.Build.targets`
----

Could you also add the above contents to `src/mono/wasm/README.md` at the end?

fanyang-mono replied:
There is a `README.md` under `src/mono/wasm/build`. I feel that it is a better location. What do you think?

fanyang-mono replied:
Updated the dependencies and added doc.

radical commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -134,7 +134,26 @@
   </Target>
 
-  <Target Name="_WasmGenerateAppBundle" Condition="'$(WasmGenerateAppBundle)' == 'true'">
+  <Target Name="WasmGenerateRuntimeConfig">
```
rename to `_WasmGenerateRuntimeConfig`

fanyang-mono replied:
Renamed.

radical commented on src/tests/FunctionalTests/WebAssembly/Browser/RuntimeConfig/Program.cs:
```diff
@@ -0,0 +1,23 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
This functional test should be added in `Wasm.Build.Tests` too, so we can test it against EMSDK, and workloads.

fanyang-mono replied:
I took a look at the `Wasm.Build.Tests`. There is only one project file. I could add a runtimeconfig.template.json file there. Then add a test case in one or some of the test files to check if it gets picked up properly. Or do you think it is better to add a brand new test file for this?

fanyang-mono replied:
Created an issue for this task. https://github.com/dotnet/runtime/issues/56799

fanyang-mono commented on src/mono/wasm/build/WasmApp.InTree.targets:
```diff
@@ -1,6 +1,9 @@
   <UsingTask TaskName="MonoAOTCompiler" AssemblyFile="$(MonoAOTCompilerTasksAssemblyPath)" />
+  <UsingTask TaskName="RuntimeConfigParserTask"
+             AssemblyFile="$(RuntimeConfigParserTasksAssemblyPath)"
+             Condition="'$(RuntimeConfigParserTasksAssemblyPath)' != ''" />
```
I would expect the task to always be there. The condition is for extra safety.

radical replied:
If it doesn't fail here, then it will fail at the call site anyway.

fanyang-mono replied:
Right, this doesn't prevent further failure caused by missing the task. Not much of use. I will remove it.

fanyang-mono commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -134,7 +134,26 @@
+    </PropertyGroup>
+
+    <ItemGroup Condition="'$(RunAOTCompilation)' == 'true'">
+      <RuntimeConfigReservedProperties Include="RUNTIME_IDENTIFIER"/>
+      <RuntimeConfigReservedProperties Include="APP_CONTEXT_BASE_DIRECTORY"/>
```
I don't understand your question here, in terms of why you would think they are AOT specific. These two properties are the ones set via `monovm_initialize` explicitly in wasm driver.c. So we don't want the user overwrites them.

radical replied:
> I don't understand your question here, in terms of why you would think they are AOT specific. 

` <ItemGroup Condition="'$(RunAOTCompilation)' == 'true'">`

I was thinking because of this condition.

Also, we should probably add a comment in `driver.c` with those keys, suggesting that this file should be updated whenever that list of keys changes.

fanyang-mono replied:
Ah I see, it is the condition for these properties. I need to look into it to confirm if the condition is needed. I didn't think too much about it when following the example for android and ios https://github.com/dotnet/runtime/blob/881f3c6c2b9800dfce3c0d2dd1a98771364754e2/eng/testing/tests.mobile.targets#L61

I will add a comment to `driver.c`

fanyang-mono replied:
Removed the AOT condition.

fanyang-mono commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -151,6 +170,9 @@
       <WasmNativeAsset Include="$(MicrosoftNetCoreAppRuntimePackRidNativeDir)$(WasmIcuDataFileName)" Condition="'$(InvariantGlobalization)' != 'true'" />
       <WasmNativeAsset Include="$(MicrosoftNetCoreAppRuntimePackRidNativeDir)dotnet.timezones.blat" />
+
+      <WasmExtraFilesToDeploy Include="$(TargetDir)publish\runtimeconfig.bin" TargetPath="managed" />
+      <WasmFilesToIncludeInFileSystem Include="$(TargetDir)publish\runtimeconfig.bin" />
```
Yes. The first one is to copy `runtimeconfig.bin` file to `AppBundle` folder. The second one is to make sure the file is visible in Wasm system.

radical replied:
> The first one is to copy runtimeconfig.bin file to AppBundle folder. 

Why is this needed? Including it in `WasmFilesToIncludeInFileSystem` should ensure that it will be copied to the app bundle.

fanyang-mono replied:
Then, line 174 is probably not needed. I added line 174 before 175.

fanyang-mono commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -134,7 +134,26 @@
+      <RuntimeConfigReservedProperties Include="APP_CONTEXT_BASE_DIRECTORY"/>
+    </ItemGroup>
+
+    <!-- Parse runtimeconfig.template.json file -->
```
Good catch! Fixed comment.

fanyang-mono commented on src/mono/wasm/runtime/driver.c:
```diff
@@ -502,6 +512,24 @@ mono_wasm_load_runtime (const char *unused, int debug_level)
 	appctx_values [1] = "browser-wasm";
 
+	char *file_name = RUNTIMECONFIG_BIN_FILE;
+    int str_len = strlen (file_name) + 1; // +1 is for the "/"
+    char *file_path = (char *)malloc (sizeof (char) * (str_len +1)); // +1 is for the terminating null character
+    int num_char = snprintf (file_path, (str_len + 1), "/%s", file_name);
+    struct stat buffer;
+
+    assert (num_char > 0 && num_char == str_len);
+
+    if (stat (file_path, &buffer) == 0) {
+        MonovmRuntimeConfigArguments *arg = (MonovmRuntimeConfigArguments *)malloc (sizeof (MonovmRuntimeConfigArguments));
+        arg->kind = 0;
+        arg->runtimeconfig.name.path = file_path;
+        monovm_runtimeconfig_initialize (arg, cleanup_runtime_config, file_path);
+    } else {
+        free (file_path);
+    }
```
Fixed.

fanyang-mono commented on src/mono/wasm/runtime/driver.c:
```diff
@@ -502,6 +512,24 @@ mono_wasm_load_runtime (const char *unused, int debug_level)
+    } else {
+        free (file_path);
+    }
+
```
Fixed.

fanyang-mono committed:
Resolve review feedback part 1

radical commented on src/tests/FunctionalTests/WebAssembly/Browser/RuntimeConfig/WebAssembly.Browser.RuntimeConfig.Test.csproj:
```diff
@@ -0,0 +1,21 @@
+  
+  <ItemGroup>
+    <Compile Include="Program.cs" />
+    <Content Include="index.html">
+      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
+    </Content>
+  </ItemGroup>
+
+  <Target Name="AfterWasmBuildApp" AfterTargets="WasmBuildApp">
+    <Copy SourceFiles="$(OutDir)\index.html" DestinationFolder="$(WasmAppDir)" />
+  </Target>
```
The `<Content..` can be dropped, since we are explicitly copying it in the `AfterWasmBuildApp` target anyway. And in the `Copy`, we can use `<Copy SourceFiles="$(MSBuildThisFileDirectory)index.html" DestinationFolder="$(WasmAppDir)" />`

fanyang-mono replied:
This is indeed necessary. Otherwise, I hit this error
```
/Users/fanyang/Documents/work/dotnet_wasm/runtime/src/tests/FunctionalTests/WebAssembly/Browser/RuntimeConfig/WebAssembly.Browser.RuntimeConfig.Test.csproj(14,5): error MSB3030: Could not copy the file "/Users/fanyang/Documents/work/dotnet_wasm/runtime/artifacts/bin/WebAssembly.Browser.RuntimeConfig.Test/net6.0-Release/browser-wasm/index.html" because it was not found.
```

radical replied:
Yep, that's why I suggested:
`<Copy SourceFiles="$(MSBuildThisFileDirectory)index.html" DestinationFolder="$(WasmAppDir)" />`

note that it uses `$(MSBuildThisFileDirectory)index.html`.

What happens here is that the `@(Content)` item will cause the file to be copied to the `$(OutDir)`. And that would end up in `publish` dir, but `WasmAppBuilder` overwrites that with it's own `index.html`. So, `AfterTargets="WasmBuildApp"`, we explicitly want to overwrite with our `index.html`.

And we don't run from `$(OutDir)`, so we don't really need `index.html` there. So, instead of going through all of that, we can directly copy `$(MSBuildThisFileDirectory)index.html` to the app bundle dir.

fanyang-mono replied:
Ah gotcha!

radical commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -134,7 +134,26 @@
+    </ItemGroup>
+
+    <!-- Parse *.runtimeconfig.json file -->
+    <RuntimeConfigParserTask
```
Corresponding to this, the `RuntimeConfigParser` task should error out if the config file does not exist.

fanyang-mono replied:
Added error checking.

fanyang-mono committed:
Resolve review feedbacks part 2

fanyang-mono committed:
Add doc for updating dependencies when adding new tasks

radical commented:
Does this PR mean that setting environment variables like [this](https://docs.microsoft.com/en-us/dotnet/core/run-time-config/#environment-variables) would work? If so, then that should be included in the Wasm.Build.Tests test being added for runtimeconfig.

radical commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -151,6 +172,8 @@
       <WasmNativeAsset Include="$(MicrosoftNetCoreAppRuntimePackRidNativeDir)$(WasmIcuDataFileName)" Condition="'$(InvariantGlobalization)' != 'true'" />
       <WasmNativeAsset Include="$(MicrosoftNetCoreAppRuntimePackRidNativeDir)dotnet.timezones.blat" />
+
+      <WasmFilesToIncludeInFileSystem Include="$(TargetDir)publish\runtimeconfig.bin" />
```
Move this to the end of `_WasmGenerateRuntimeConfig` target, since it belongs there. And it gets added only if we run that target.

fanyang-mono replied:
Make sense!

radical commented:
Not for this PR, but we maybe need to handle the case when you have
```json
{
   "runtimeOptions": {
      "configProperties": {
         "System.Globalization.Invariant": true
      }
   }
}
```

.. in your `runtimeconfig.template.json`, then the wasm targets should treat this as `$(InvariantGlobalization)=true`. @lewing ?

fanyang-mono committed:
Remove redundant condition and redundant copy

fanyang-mono commented:
> Not for this PR, but we maybe need to handle the case when you have
> 
> ```json
> {
>    "runtimeOptions": {
>       "configProperties": {
>          "System.Globalization.Invariant": true
>       }
>    }
> }
> ```
> 
> .. in your `runtimeconfig.template.json`, then the wasm targets should treat this as `$(InvariantGlobalization)=true`. @lewing ?

My `runtimeconfig.template.json` is only valid for the test I added. It needs to be added for the app who needs the configuration of `$(InvariantGlobalization)=true`. To my understanding, `runtimeconfig.template.json` needs to be visible during the build/publish process.

lambdageek commented:
> Not for this PR, but we maybe need to handle the case when you have
> 
> ```json
> {
>    "runtimeOptions": {
>       "configProperties": {
>          "System.Globalization.Invariant": true
>       }
>    }
> }
> ```
> 
> .. in your `runtimeconfig.template.json`, then the wasm targets should treat this as `$(InvariantGlobalization)=true`. @lewing ?

This doesn't make a lot of sense to me - `runtime.template.json` is meant to configure the runtime behavior of the app, not how it is built/packaged.  Reading it at build-time doesn't make a lot of sense to me.

fanyang-mono committed:
There is no runtimeconfig.json file to parse, when OutputType is library. So skip it.

fanyang-mono committed:
Skip for Android tests as well.

radical commented:
> `runtime.template.json` is meant to configure the runtime behavior of the app, not how it is built/packaged. Using it to configure the build doesn't make a lot of sense to me.

What happens if you only set the `"System.Globalization.Invariant": true` in your `runtimeconfig.template.json`, in a non-wasm project? Is that the equivalent to using `/p:InvariantGlobalization=true`?

If so, then in case of wasm, we need to link out relevant bits, IIUC, and for that wasm targets would need to read the config file.

fanyang-mono committed:
Remove the wrong locatioin of WasmFilesToIncludeInFileSystem

fanyang-mono committed:
Update runtimeconfig file path

fanyang-mono committed:
Add RuntimeConfigParserTasksAssemblyPath to linker test project template

radical commented on eng/testing/linker/trimmingTests.targets:
```diff
@@ -96,6 +96,7 @@
                                                  .Replace('{MonoProjectRoot}', '$(MonoProjectRoot)')
                                                  .Replace('{MonoAOTCompilerTasksAssemblyPath}', '$(MonoAOTCompilerTasksAssemblyPath)')
                                                  .Replace('{JsonToItemsTaskFactoryTasksAssemblyPath}', '$(JsonToItemsTaskFactoryTasksAssemblyPath)')
+                                                 .Replace('{RuntimeConfigParserTasksAssemblyPath}', '$(RuntimeConfigParserTasksAssemblyPath)')
```
Similar line needed for `RuntimeConfigParserTaskDir`?

fanyang-mono replied:
CI seems to be working well without it so far. If there is a problem, I will add it.

radical commented on eng/testing/tests.mobile.targets:
```diff
@@ -52,13 +52,13 @@
              Condition="'$(RuntimeConfigParserTasksAssemblyPath)' != ''" />
 
-  <Target Name="GenerateRuntimeConfig">
+  <Target Name="GenerateRuntimeConfig" Condition="'$(OutputType)' != 'library'">
```
This shouldn't be needed for browser, correct? If so, then can we add !browser to the condition? I know the target is being used for non-browser cases, but adding to the condition would prevent it from getting used by mistake.

fanyang-mono replied:
Added condition.

radical commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -134,7 +134,33 @@
   </Target>
 
-  <Target Name="_WasmGenerateAppBundle" Condition="'$(WasmGenerateAppBundle)' == 'true'">
+  <Target Name="_WasmGenerateRuntimeConfig" Condition="'$(OutputType)' != 'library'">
```
Where did it hit the `library` case?

radical replied:
And why was `'$(WasmGenerateAppBundle)' == 'true'` removed?

fanyang-mono replied:
> Where did it hit the `library` case?

https://github.com/dotnet/runtime/blob/main/src/tests/FunctionalTests/WebAssembly/Browser/HotReload/ApplyUpdateReferencedAssembly/ApplyUpdateReferencedAssembly.csproj#L5

fanyang-mono replied:
> And why was `'$(WasmGenerateAppBundle)' == 'true'` removed?

That has always been there for another target `_WasmGenerateAppBundle`, not the one I added. The diff is misleading there.

radical replied:
oops, I misread that!

lewing commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -134,7 +134,33 @@
+    </ItemGroup>
+  </Target>
+
+  <Target Name="_WasmGenerateAppBundle" Condition="'$(WasmGenerateAppBundle)' == 'true'" DependsOnTargets="_WasmGenerateRuntimeConfig">
```
If this is intended to work in Blazor this doesn't look sufficient since _WasmGenerateAppBundle does not run in that case.

fanyang-mono replied:
Where is `WasmAppBuilder` called for Blazor?

radical replied:
Blazor doesn't use `WasmAppBuilder`. It has its own packaging/bundling stuff.

fanyang-mono replied:
Where does that logic live?

fanyang-mono replied:
Created an issue to track this. https://github.com/dotnet/aspnetcore/issues/35010

lewing replied:
There is a broader problem with making this work in blazor in that it requires a workload and the baseline blazor scenarios are supposed to be functional without any workloads installed.

fanyang-mono replied:
Created a tracking issue https://github.com/dotnet/aspnetcore/issues/35010

lewing reviewed (commented):
This is a good start but doesn't look like it will work in blazor apps

fanyang-mono committed:
Fix test failures on WASM EAT and AOT lanes

fanyang-mono commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -134,7 +134,26 @@
+  <Target Name="WasmGenerateRuntimeConfig">
+    <PropertyGroup>
+      <RuntimeConfigFilePath>$(TargetDir)publish\$(AssemblyName).runtimeconfig.json</RuntimeConfigFilePath>
+      <ParsedRuntimeConfigFilePath>$(TargetDir)publish\runtimeconfig.bin</ParsedRuntimeConfigFilePath>
```
Fixed.

fanyang-mono commented:
Lane `dotnet-linker-tests (Build windows x64 release Runtime_Release)` actually passed for rerun. Don't know why it still showed up as failed. https://dev.azure.com/dnceng/public/_build/results?buildId=1272100&view=results

fanyang-mono commented:
@radical @lewing If this change looks good to you, could you please approve it?

radical committed:
[wasm] Add a test for reading runtimeconfig.json

radical committed:
improve error reporting slightly

radical commented:
There are some disabled tests with ActiveIssue on https://github.com/dotnet/runtime/issues/38433 being fixed here, which should be enabled in this PR.

fanyang-mono committed:
Enable tests

fanyang-mono committed:
One more thing to enable tests

fanyang-mono committed:
Run target _WasmGenerateRuntimeConfig when runtimeconfig.json file exists

fanyang-mono committed:
Add a test for configure runtime using msbuild properties

fanyang-mono committed:
Fix runtime test failures

radical commented on src/tests/BuildWasmApps/Wasm.Build.Tests/WasmBuildAppTest.cs:
```diff
@@ -85,6 +85,89 @@ public void Bug49588_RegressionTest_NativeRelinking(BuildArgs buildArgs, RunHost
+        public void PropertiesFromCsproj(BuildArgs buildArgs, RunHost host, string id)
+        {
+            buildArgs = buildArgs with { ProjectName = $"runtime_config_csproj_{buildArgs.Config}_{buildArgs.AOT}" };
+            buildArgs = ExpandBuildArgs(buildArgs);
```
Use `ExpandBuildArgs(buildArgs, extraProperties: "<ThreadPoolMaxThreads>20</ThreadPoolMaxThreads>");`

fanyang-mono replied:
updated the test.

fanyang-mono committed:
Update tests base on feedback

fanyang-mono committed:
Add missing ")"

radical commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -122,6 +124,10 @@
 
       <WasmAppDir>$([MSBuild]::NormalizeDirectory($(WasmAppDir)))</WasmAppDir>
+
+      <_MainAssemblyPath Condition="'%(WasmAssembliesToBundle.FileName)' == $(AssemblyName) and '%(WasmAssembliesToBundle.Extension)' == '.dll'">%(WasmAssembliesToBundle.Identity)</_MainAssemblyPath>
```
We should do this if `$(WasmGenerateAppBundle) == 'true'`. If the assembly is not found, then we should emit a warning.

fanyang-mono replied:
Ah, that's the condition that I was missing...

radical reviewed (changes requested):
Looks good, just couple of comments

fanyang-mono commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -83,6 +83,8 @@
 
     <_BeforeWasmBuildAppDependsOn />
+    
+    <_RuntimeConfigJsonFileExistance>false</_RuntimeConfigJsonFileExistance>
```
I thought that it is cleaner to do it this way.

radical replied:
Exists check on the target is idiomatic for msbuild.

fanyang-mono committed:
Add one more condition

fanyang-mono committed:
More format changes

fanyang-mono committed:
Better error handling

radical committed:
[wasm] Fix build for a test support project

This fixes the build for ApplyUpdateReferencedAssembly.csproj, which is
a support project for HotReload wasm functional test.

```
D:\workspace\_work\1\s\src\mono\wasm\build\WasmApp.targets(131,5): error : Could not find
    D:\workspace\_work\1\s\artifacts\bin\ApplyUpdateReferencedAssembly\net6.0-Release\browser-wasm\publish\ApplyUpdateReferencedAssembly.runtimeconfig.json
    for D:\workspace\_work\1\s\artifacts\bin\ApplyUpdateReferencedAssembly\net6.0-Release\browser-wasm\publish\ApplyUpdateReferencedAssembly.dll.

    [D:\workspace\_work\1\s\src\tests\FunctionalTests\WebAssembly\Browser\HotReload\ApplyUpdateReferencedAssembly\ApplyUpdateReferencedAssembly.csproj]
```

The wasm targets should be run at all for this project. But they are run
because they get imported by tests.wasm.targets, which gets imported because
`$(IsTestProject)=true`.

The csproj has `$(IsTestProject)=false`, and `$(IsTestSupportProject)=true`,
which should mean that the test.props/targets don't get imported. But the
import is conditioned on `$(EnableTestSupport)`, which gets set in
`src/libraries/Directory.Build.props`. And that means setting
`$(IsTestProject)=true` in the csproj is too late.

So, instead, set that in a `Directory.Build.props`. And also, ensure that the
`Directory.Build.props` for functional tests doesn't override the value!

radical reviewed (approved):
LGTM, thanks for your patience! 👍
